### PR TITLE
Hsm Authentication with Http over Unix domain sockets

### PR DESCRIPTION
**Pull request by ancaantochi**
# Description of the changes
Added support for HSM authentication calls over Unix Domain Sockets

### CODE REVIEW

File: iothub/device/src/HsmAuthentication/HttpHsmSignatureProvider.cs [hunk 1]
```diff
@@ -2,19 +2,27 @@
...
 namespace Microsoft.Azure.Devices.Client.HsmAuthentication
 {
     class HttpHsmSignatureProvider : ISignatureProvider
     {
-        const string DefaultApiVersion = "2018-06-28";
-        const SignRequestAlgo DefaultSignRequestAlgo = SignRequestAlgo.HMACSHA256;
-        const string DefaultKeyId = "primary";
-        readonly string apiVersion;
-        readonly HsmHttpClient httpClient;
+        private const string DefaultApiVersion = "2018-06-28";
```
Comment (zolvarga) [inline iothub/device/src/HsmAuthentication/HttpHsmSignatureProvider.cs]: >2018-06-28 [](start = 50, length = 10)

Is this API version is not affected by the pre GA post GA version changes?

File: iothub/device/src/HsmAuthentication/GeneratedCode/HttpHsmClient.cs [hunk 2]
```diff
@@ -152,8 +154,240 @@ public System.Threading.Tasks.Task<SignResponse> SignAsync(string api_version, s
...
+                            var responseData_ = response_.Content == null ? null : await response_.Content.ReadAsStringAsync().ConfigureAwait(false);
+                            var result_ = default(EncryptResponse);
+                            try
+                            {
+                                result_ = Newtonsoft.Json.JsonConvert.DeserializeObject<EncryptResponse>(responseData_, _settings.Value);
+                                return result_;
+                            }
+                            catch (System.Exception exception_)
+                            {
+                                throw new SwaggerException("Could not deserialize the response body.", (int)response_.StatusCode, responseData_, headers_, exception_);
```
Comment (zolvarga) [inline iothub/device/src/HsmAuthentication/GeneratedCode/HttpHsmClient.cs]: >SwaggerException [](start = 42, length = 16)

I don't see how the Swagger class involved here...

File: iothub/device/src/netstandard20/HsmAuthentication/Transport/HttpBufferedStream.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.HsmAuthentication.Transport
+{
+    class HttpBufferedStream : Stream
+    {
+        private const char CR = '\r';
+        private const char LF = '\n';
+        private BufferedStream innerStream;
```
Comment (zolvarga) [inline iothub/device/src/netstandard20/HsmAuthentication/Transport/HttpBufferedStream.cs]: >innerStream [](start = 31, length = 11)

Coding standard: use underscore prefix for private members

File: iothub/device/src/netstandard20/HsmAuthentication/Transport/HttpUdsMessageHandler.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+using System.Net.Http;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.HsmAuthentication.Transport
+{
+    internal class HttpUdsMessageHandler : HttpMessageHandler
+    {
+        private readonly Uri providerUri;
```
Comment (zolvarga) [inline iothub/device/src/netstandard20/HsmAuthentication/Transport/HttpUdsMessageHandler.cs]: >providerUri [](start = 29, length = 11)

[https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)

File: iothub/device/src/HsmAuthentication/GeneratedCode/HttpHsmClient.cs (hunk 2 above)
Comment (ancaantochi) [reply iothub/device/src/HsmAuthentication/GeneratedCode/HttpHsmClient.cs]: This class is generated by Swagger. Calling methods catch SwaggerExceptions and throw HttpHsmComunicationException

File: iothub/device/src/HsmAuthentication/HttpHsmSignatureProvider.cs (hunk 1 above)
Comment (ancaantochi) [reply iothub/device/src/HsmAuthentication/HttpHsmSignatureProvider.cs]: This is the GA version, pre GA we will use the same version and post GA if there are changes will have to publish a new package that uses the new version. I think this is similar to sdk version for IotHub.

**Pull request merged**